File: py110/small_problems_110_119/medium2/1.py
# https://launchschool.com/exercises/5800a75a?track=python

'''

input: string
output: dictionary - 
    1. % of chars in the string that are lowercaase
    2. % of chars that are upper
    3. % of chars that are neither


rules:
1. % are strings whose numeric values are between 0.00 and 100.00
2. Rounded to two decimal places
3. Assume at least one char.
4. spaces are counted as neither

Algorithm:
1. find length of string
2. create 3 buckets - lower, upper, and neither
3. Iterate through the string and depending on condiiton add to bucket
4. Create dictionary with the buckets as values and the keys of lower, upper, and neither
5. round to 2 decimal places


'''
def convert_to_percentage(number, length):
    return (number / length) * 100


def letter_percentages(string):
    length = len(string)
    lower = 0
    upper = 0
    neither = 0

    for char in string:
        if char.isupper():
            upper += 1
        elif char.islower():
            lower += 1
        else:
            neither += 1
    

    return {"lowercase": f"{convert_to_percentage(lower, length):.2f}", 
            "uppercase": f"{convert_to_percentage(upper, length):.2f}",
             "neither": f"{convert_to_percentage(neither, length):.2f}"}

expected_result = {
    'lowercase': "50.00",
    'uppercase': "10.00",
    'neither': "40.00",
}
print(letter_percentages('abCdef 123') == expected_result)

expected_result = {
    'lowercase': "37.50",
    'uppercase': "37.50",
    'neither': "25.00",
}
print(letter_percentages('AbCd +Ef') == expected_result)

expected_result = {
    'lowercase': "0.00",
    'uppercase': "0.00",
    'neither': "100.00",
}
print(letter_percentages('123') == expected_result)

File: ls215/smallProblems/med2/1.js
'use strict';

/*
input: string
output: object containing three properties: lowercase, uppercase, neither
rules:
1. % of characters in the string that are lowercase letter
2. % of characters that are uppercase letter
3. % of characters that are neither.
4. string will always contain at least one char.


abCdef 123 -> 10
1 / 10 -> 10%
5 /10 -> 50


AbCd +Ef -> 8
 3/ 8 -> 37.50
 3/ 8 -> 37.50
 2 /8 -> 25

DS:


Algo:
1. Create object with 3 properties to hold the amount of chars
1a. find length of string
2. Iterate through string
3. If char is uppercase add it to prop upper
3a. if lower cadd to prop lower
3c. Anything else add to prop neither
4. return object with 3 props and their total over the total length.
4a. format to two decimal places

*/



// function letterPercentages(string) {
//   let objCount = {'lowercase': 0, 'uppercase': 0, 'neither': 0};
//   let length = string.length;

//   for (let char of string) {
//     if (/[A-Z]/.test(char)) {
//       objCount.uppercase += 1;
//     } else if (/[a-z]/.test(char)) {
//       objCount.lowercase += 1;
//     } else {
//       objCount.neither += 1;
//     }
//   }

//   return {
//     lowercase: formatString(objCount.lowercase, length),
//     uppercase: formatString(objCount.uppercase, length),
//     neither: formatString(objCount.neither, length),
//   };
// }

// function formatString(count, length) {
//   return String((count / length * 100).toFixed(2));
// }


function letterPercentages(string) {
  let count = string.length;

  return {
    lowercase: (((string.match(/[a-z]/g) || []).length / count) * 100).toFixed(2),
    uppercase: (((string.match(/[A-Z]/g) || []).length / count) * 100).toFixed(2),
    neither: (((string.match(/[^a-z]/ig) || []).length / count) * 100).toFixed(2),
  };
}


console.log(letterPercentages('abCdef 123'));
// { lowercase: "50.00", uppercase: "10.00", neither: "40.00" }

console.log(letterPercentages('AbCd +Ef'));
// { lowercase: "37.50", uppercase: "37.50", neither: "25.00" }

console.log(letterPercentages('123'));
// { lowercase: "0.00", uppercase: "0.00", neither: "100.00" }